### PR TITLE
Update IStack.h

### PR DESCRIPTION
**Pull request by flakeed**

### CODE REVIEW

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+namespace Platform::Collections::Stacks
+{
+    template<typename TElement>
+    concept CStack = requires(TElement item)
+    {
+        { self.empty() } -> std::same_as<bool>;
+
+        { self.push(item) };
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/CStack.h:10]: https://en.cppreference.com/w/cpp/container/stack/push

File: cpp/Platform.Collections/Stacks/IStack.h [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
+    template <typename ...>
+    struct IStack;
+
+    template<typename TElement>
+    struct IStack<TElement>
     {
-        { self.empty() } -> std::same_as<bool>;
+        virtual bool empty() = 0;
+
+        virtual void push(TElement item) = 0;
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/IStack.h:13]: https://en.cppreference.com/w/cpp/container/stack/push

**Review by Konard — CHANGES_REQUESTED** [review]

File: cpp/Platform.Collections/Stacks/CStack.h (hunk 1 above)
Comment (uselessgoddess) [reply cpp/Platform.Collections/Stacks/CStack.h:10]: What, bro?

File: cpp/Platform.Collections/Stacks/IStack.h (hunk 2 above)
Comment (uselessgoddess) [reply cpp/Platform.Collections/Stacks/IStack.h:13]: Also, bro

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
+namespace Platform::Collections::Tests {
+struct EmptyStack : public Stacks::IStack<int> {
+  bool empty() const {}
```
Comment (Konard) [inline cpp/Platform.Collections.Tests/StackTests.cpp:3]: Please use 4 space indentation in the whole file.

File: cpp/Platform.Collections/Platform.Collections.h [hunk 4]
```diff
@@ -11,6 +11,8 @@
 #include <span>
 #include <iostream>
 #include <numeric>
+#include <stack>./cpp/build_and_test_gitpod.sh
```
Comment (Konard) [inline cpp/Platform.Collections/Platform.Collections.h:14]:
```suggestion
#include <stack>
```

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  const int &top() const {}
+};
+TEST(StackTests, Concept) {
+  Stacks::CStack<int> auto stack = std::stack<int>{};
+  stack.push(1);
+  ASSERT_EQ(stack.top(), 1);
+  stack.pop();
+  ASSERT_TRUE(stack.empty());
+}
+TEST(StackTests, Interface) {}
```
Comment (Konard) [inline cpp/Platform.Collections.Tests/StackTests.cpp:20]: Why do we need this empty test?

File: cpp/Platform.Collections/Platform.Collections.h (hunk 4 above)
Comment (Konard) [reply cpp/Platform.Collections/Platform.Collections.h:14]: But I’m not sure we need to include stack header into our library by default. If it is needed in test it should be included there.

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        requires(TSelf& self,TElement item) 
+        {
+            { self.empty() } -> std::same_as<bool>;
+
+            { self.push(item) };
+
+            { self.pop() };
+
+            { self.top() } -> std::same_as<TElement&>;
+        }and
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/CStack.h:16]:
```suggestion
        } &&
```

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+namespace Platform::Collections::Stacks
+{
+    template<typename TSelf,typename TElement>
+    concept CStack =
+        requires(TSelf& self,TElement item)
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/CStack.h:7]:
```suggestion
        requires(TSelf& self, TElement item) 
```

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            { self.empty() } -> std::same_as<bool>;
+
+            { self.push(item) };
+
+            { self.pop() };
+
+            { self.top() } -> std::same_as<TElement&>;
+        }and
+        requires(const TSelf& self,TElement item)
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/CStack.h:17]:
```suggestion
        requires(const TSelf& self, TElement item)
```

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+namespace Platform::Collections::Stacks
+{
+    template<typename TSelf,typename TElement>
+    concept CStack =
+        requires(TSelf& self,TElement item) 
+        {
+            { self.empty() } -> std::same_as<bool>;
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/CStack.h:9]: Does `std::stack` has a non const `empty` method?

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+            { self.push(item) };
+
+            { self.pop() };
+
+            { self.top() } -> std::same_as<TElement&>;
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/CStack.h:15]: Same question here.

**Review by Konard — CHANGES_REQUESTED** [review]

File: cpp/Platform.Collections/Stacks/CStack.h (hunk 6 above)
Comment (uselessgoddess) [reply cpp/Platform.Collections/Stacks/CStack.h:16]:
```suggestion
        } and
```

File: cpp/Platform.Collections/Stacks/CStack.h (hunk 9 above)
Comment (uselessgoddess) [reply cpp/Platform.Collections/Stacks/CStack.h:9]: Remove it

File: cpp/Platform.Collections/Stacks/CStack.h (hunk 10 above)
Comment (uselessgoddess) [reply cpp/Platform.Collections/Stacks/CStack.h:15]: Not same. `TElement` reference is mutable

File: cpp/Platform.Collections.Tests/StackTests.cpp (hunk 5 above)
Comment (uselessgoddess) [reply cpp/Platform.Collections.Tests/StackTests.cpp:20]: To write: `Stacks::CStack<int> auto stack = EmptyStack;`

Comment (Konard) [reply cpp/Platform.Collections.Tests/StackTests.cpp:20]: Are you sure? We have this code statement in the next test.

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+{
+    template<typename TSelf, typename TElement>
+    concept CStack =
+        requires(TSelf & self, TElement item)
+    {
+        { self.push(item) };
+
+        { self.pop() };
+
+        { self.top() } -> std::same_as<TElement&>;
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/CStack.h:13]: Do we need check this method here? Isn’t it enough to just check the const version below?

File: cpp/Platform.Collections/Stacks/IStack.h [hunk 12]
```diff
@@ -1,14 +1,21 @@
...
     {
-        { self.empty() } -> std::same_as<bool>;
+        virtual bool empty() const = 0;
+
+        virtual void push(TElement item) = 0;
+
+        virtual void pop() = 0;
 
-        { self.push(item) };
+        virtual TElement& top() = 0;
```
Comment (Konard) [inline cpp/Platform.Collections/Stacks/IStack.h:15]: Do we need to support this version of `top` method in interface?

**Review by Konard — CHANGES_REQUESTED** [review]

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
```
Comment (uselessgoddess) [inline cpp/Platform.Collections.Tests/StackTests.cpp:20]:
```suggestion
    Stacks::CStack<int> auto stack = std::stack<int>{};
    stack.push(1);
    ASSERT_EQ(stack.top(), 1);
    stack.pop();
    ASSERT_TRUE(stack.empty());
}

TEST(StackTests, Interface) {}
```

File: cpp/Platform.Collections/Platform.Collections.h (hunk 4 above)
Comment (uselessgoddess) [inline cpp/Platform.Collections/Platform.Collections.h:14]:
```suggestion
```

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    Stacks::CStack<int> auto stack = std::stack<int>{};
+    stack.push(1);
+    ASSERT_EQ(stack.top(), 1);
+    stack.pop();
+    ASSERT_TRUE(stack.empty());
+}
+
+TEST(StackTests, Interface) {}
+TEST(StackTests, ConceptInterface) {}
+} // namespace Platform::Collections::Tests
```
Comment (uselessgoddess) [inline cpp/Platform.Collections.Tests/StackTests.cpp:29]:
```suggestion
}
```

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+{
+    Stacks::CStack<int> auto stack = std::stack<int>{};
+    stack.push(1);
+    ASSERT_EQ(stack.top(), 1);
+    stack.pop();
+    ASSERT_TRUE(stack.empty());
+}
+
+TEST(StackTests, Interface) {}
+TEST(StackTests, ConceptInterface) {}
```
Comment (uselessgoddess) [inline cpp/Platform.Collections.Tests/StackTests.cpp:28]:
```suggestion
    struct EmptyStack : public Stacks::IStack<int> 
    {
        bool empty() const {}

        void push(int item) {}

        void pop() {}

        int &top() {}

        const int &top() const {}
    };

    TEST(StackTests, Concept) 
    {
        Stacks::CStack<int> auto stack = std::stack<int>{};
        stack.push(1);
        ASSERT_EQ(stack.top(), 1);
        stack.pop();
        ASSERT_TRUE(stack.empty());
    }

    TEST(StackTests, Interface) {}
    TEST(StackTests, ConceptInterface) {}
```

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    TEST(StackTests, Concept) 
+    {
+        Stacks::CStack<int> auto stack = std::stack<int>{};
+        stack.push(1);
+        ASSERT_EQ(stack.top(), 1);
+        stack.pop();
+        ASSERT_TRUE(stack.empty());
+    }
+
+    TEST(StackTests, Interface) {}
```
Comment (uselessgoddess) [inline cpp/Platform.Collections.Tests/StackTests.cpp:27]:
```suggestion
    TEST(StackTests, Interface) 
    {
        std::unique_ptr<Stack::IStack<int>> stack = std::make_unique<EmptyStack>();
    }
```

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    TEST(StackTests, ConceptInterface) {}
```
Comment (uselessgoddess) [inline cpp/Platform.Collections.Tests/StackTests.cpp:28]:
```suggestion
    TEST(StackTests, ConceptInterface) 
    {
        Stacks::CStack<int> auto stack  = EmptyStack{};
    }
```

File: cpp/Platform.Collections.Tests/StackTests.cpp [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        stack.push(1);
+        ASSERT_EQ(stack.top(), 1);
+        stack.pop();
+        ASSERT_TRUE(stack.empty());
+    }
+
+    TEST(StackTests, Interface) 
+    {
+        std::unique_ptr<Stack::IStack<int>> stack = std::make_unique<EmptyStack>();
+    }
```
Comment (uselessgoddess) [inline cpp/Platform.Collections.Tests/StackTests.cpp:30]:
```suggestion
    }
    
```

File: cpp/Platform.Collections/Stacks/CStack.h [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+        { self.push(item) };
+    } &&
+        requires(const TSelf& self, TElement item)
+    {
+        { self.empty() } -> std::same_as<bool>;
+
+        { self.pop() };
+
+        { self.top() } -> std::same_as<TElement&>;
+    };
```
Comment (uselessgoddess) [inline cpp/Platform.Collections/Stacks/CStack.h:18]:
```suggestion
    concept CStack = requires(TSelf& self, TElement item)
    {
        { self.push(item) };
    } 
    &&
    requires(const TSelf& self, TElement item)
    {
        { self.empty() } -> std::same_as<bool>;

        { self.pop() };

        { self.top() } -> std::same_as<TElement&>;
    };
```

**Pull request merged**